core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -106,10 +106,16 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
             "Missing required field: %s", field.name());
         // Create a field that will be defaulted to null. We assign a unique suffix to the field
         // to make sure that even if records in the file have the field it is not projected.
+        String origFieldName = field.name();
+        boolean isValidFieldName = AvroSchemaUtil.validAvroName(origFieldName);
+        String fieldName = isValidFieldName ? origFieldName : AvroSchemaUtil.sanitize(origFieldName);
         Schema.Field newField = new Schema.Field(
-            field.name() + "_r" + field.fieldId(),
+            fieldName + "_r" + field.fieldId(),
             AvroSchemaUtil.toOption(AvroSchemaUtil.convert(field.type())), null, JsonProperties.NULL_VALUE);
         newField.addProp(AvroSchemaUtil.FIELD_ID_PROP, field.fieldId());
+        if (!isValidFieldName) {
```

Comment:
@RussellSpitzer @chenjunjiedada We need to determine whether the original field name prop needs to be set. This same as `BuildAvroProjection`.

Reply:
We set according to comparing target `fieldName` and `originFiledName`, right? Also, can we always set the property to `originFieldName`?